File: src/use-cases/party/DeleteParty.js
/**
 * DeleteParty Use Case
 * Allows a user to delete a party if they are the owner or the only human player
 */

const logger = require('../../../logger');

class DeleteParty {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(partyRepository, userRepository) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Delete request
     * @param {string} request.userId - User ID attempting to delete
     * @param {string} request.partyId - Party ID to delete
     * @returns {Promise<Object>} Delete result
     */
    async execute({ userId, partyId }) {
        try {
            // Validate input
            if (!userId || typeof userId !== 'string') {
                throw new Error('User ID is required');
            }

            if (!partyId || typeof partyId !== 'string') {
                throw new Error('Party ID is required');
            }

            // Verify user exists
            const user = await this.userRepository.findById(userId);
            if (!user) {
                throw new Error('User not found');
            }

            // Find party
            const party = await this.partyRepository.findById(partyId);
            if (!party) {
                throw new Error('Party not found');
            }

            // Get all players in the party
            const players = await this.partyRepository.getPartyPlayers(partyId);

            // Check if user is in the party
            const userPlayer = players.find(p => p.userId === userId);
            if (!userPlayer) {
                throw new Error('User is not in this party');
            }

            // Check authorization: user must be owner OR the only human player
            const isOwner = party.ownerId === userId;

            // Count human players in the party
            let humanPlayers = [];
            for (const player of players) {
                const playerUser = await this.userRepository.findById(player.userId);
                if (playerUser && playerUser.isHuman()) {
                    humanPlayers.push(player);
                }
            }

            const isOnlyHuman = humanPlayers.length === 1 && humanPlayers[0].userId === userId;

            if (!isOwner && !isOnlyHuman) {
                throw new Error('Only the party owner or the only human player can delete the party');
            }

            // Cannot delete during active game (unless forced by being only human)
            if (party.status === 'playing' && !isOnlyHuman) {
                throw new Error('Cannot delete party during active game');
            }

            // Delete the party (cascade deletes handle players, rounds, game state)
            await this.partyRepository.delete(partyId);

            logger.info('Party deleted', {
                partyId: partyId,
                partyName: party.name,
                deletedBy: userId,
                deletedByUsername: user.username,
                wasOwner: isOwner,
                wasOnlyHuman: isOnlyHuman
            });

            return {
                success: true,
                deletedPartyId: partyId,
                deletedPartyName: party.name
            };
        } catch (error) {
            logger.error('Delete party error', {
                userId,
                partyId,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = DeleteParty;

File: zapzap-rust/src/application/party/delete_party.rs
use std::sync::Arc;

use crate::domain::entities::PartyStatus;
use crate::domain::repositories::{PartyRepository, RepositoryError, UserRepository};

/// Delete party input
pub struct DeletePartyInput {
    pub user_id: String,
    pub party_id: String,
}

/// Delete party output
pub struct DeletePartyOutput {
    pub deleted_party_id: String,
    pub deleted_party_name: String,
    /// `public` or `private`: SSE sends a public party's deletion to every stream
    pub deleted_party_visibility: String,
}

/// Delete party use case
pub struct DeleteParty<U: UserRepository, P: PartyRepository> {
    user_repo: Arc<U>,
    party_repo: Arc<P>,
}

impl<U: UserRepository, P: PartyRepository> DeleteParty<U, P> {
    pub fn new(user_repo: Arc<U>, party_repo: Arc<P>) -> Self {
        Self {
            user_repo,
            party_repo,
        }
    }

    pub async fn execute(
        &self,
        input: DeletePartyInput,
    ) -> Result<DeletePartyOutput, DeletePartyError> {
        // Find party
        let party = self
            .party_repo
            .find_by_id(&input.party_id)
            .await?
            .ok_or(DeletePartyError::PartyNotFound)?;

        // The Node backend's order: a member, then the owner or the only human
        // player, then a party not in play
        let players = self.party_repo.get_party_players(&input.party_id).await?;
        if !players.iter().any(|p| p.user_id == input.user_id) {
            return Err(DeletePartyError::NotInParty);
        }

        let mut is_human = false;
        let mut other_humans = 0;
        for p in &players {
            let Some(u) = self.user_repo.find_by_id(&p.user_id).await? else {
                continue;
            };
            if u.is_bot() {
                continue;
            }
            if p.user_id == input.user_id {
                is_human = true;
            } else {
                other_humans += 1;
            }
        }
        let is_only_human = is_human && other_humans == 0;

        if party.owner_id != input.user_id && !is_only_human {
            return Err(DeletePartyError::NotOwner);
        }

        // The only human in a game against bots may end it; anyone else waits for its end
        if party.status == PartyStatus::Playing && !is_only_human {
            return Err(DeletePartyError::PartyInProgress);
        }

        let party_id = party.id.clone();
        let party_name = party.name.clone();
        let deleted_party_visibility = party.visibility.as_str().to_string();

        // Delete party
        self.party_repo.delete(&input.party_id).await?;

        Ok(DeletePartyOutput {
            deleted_party_id: party_id,
            deleted_party_name: party_name,
            deleted_party_visibility,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum DeletePartyError {
    #[error("Party not found")]
    PartyNotFound,
    #[error("User is not in this party")]
    NotInParty,
    #[error("Not the party owner")]
    NotOwner,
    #[error("Cannot delete party in progress")]
    PartyInProgress,
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
